src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java

```diff
@@ -2120,6 +2120,35 @@ public void test_submodule_checkout_simple() throws Exception {
         assertSubmoduleContents(w.repo);
     }
 
+    /* Opening a git repository in a directory with a symbolic git file instead
+     * of a git directory should function properly. JGit does not implement
+     * renamed submodules, so we the submodules branch won't work with it...
+     */
+    @NotImplementedInJGit
+    public void test_with_repository_works_with_submodule() throws Exception {
+        w = clone(localMirror());
+        assertSubmoduleDirs(w.repo, false, false);
+
+        /* Checkout a branch which includes submodules (in modules directory) */
+        String subBranch = "tests/getSubmodules";
```

Comment:
I'm exploring further and I think that the preceding test (`test_submodule_checkout_simple`) can be extended to cover both CLI git and JGit and your case.  It appears to be missing a call to submoduleInit() in order to work with JGit and the tests/getSubmodules-jgit branch.